  @Test  
  public void testServerErrors() throws Exception {
    // tests the behaviour of the client when the server sends an error
    // must use BasicClientConnectionManager to test whether the client is closed correctly
    BasicHttpClientConnectionManager conMgr = new BasicHttpClientConnectionManager();
    Replicator replicator = new HttpReplicator(host, port, ReplicationService.REPLICATION_CONTEXT + "/s1", conMgr);
    ReplicationClient client = new ReplicationClient(replicator, new IndexReplicationHandler(handlerIndexDir, null), 
        new PerSessionDirectoryFactory(clientWorkDir));
    
    try {
      publishRevision(5);

      replicationServlet.setRespondWithError(true);
      expectThrows(Exception.class, client::updateNow);
      
      replicationServlet.setRespondWithError(false);
      client.updateNow(); // now it should work
      reopenReader();
      assertEquals(5, Integer.parseInt(reader.getIndexCommit().getUserData().get("ID"), 16));
      
      client.close();
    } finally {
      replicationServlet.setRespondWithError(false);
    }
  }

